client/src/client.rs

```diff
@@ -50,21 +49,77 @@ impl<R> Default for QueryResponseHandler<R> {
     }
 }
 
+/// `Result` with [`ClientQueryError`] as an error
+pub type QueryHandlerResult<T> = core::result::Result<T, ClientQueryError>;
+
 impl<R> ResponseHandler for QueryResponseHandler<R>
 where
     R: Query + Into<QueryBox> + Debug,
     <R::Output as TryFrom<Value>>::Error: Into<eyre::Error>,
 {
-    type Output = PaginatedQueryOutput<R>;
+    type Output = QueryHandlerResult<ClientQueryOutput<R>>;
+
+    fn handle(self, resp: Response<Vec<u8>>) -> Self::Output {
+        // Separate-compilation friendly response handling
+        fn _handle_query_response_base(
+            resp: &Response<Vec<u8>>,
+        ) -> QueryHandlerResult<VersionedPaginatedQueryResult> {
+            match resp.status() {
+                StatusCode::OK => VersionedPaginatedQueryResult::decode_versioned(resp.body())
+                    .wrap_err("Failed to decode response body as VersionedPaginatedQueryResult")
+                    .map_err(Into::into),
+                StatusCode::BAD_REQUEST
+                | StatusCode::UNAUTHORIZED
+                | StatusCode::FORBIDDEN
+                | StatusCode::NOT_FOUND => {
+                    let err = QueryError::decode(&mut resp.body().as_ref())
+                        .wrap_err("Failed to decode response body as QueryError")?;
+                    Err(ClientQueryError::Certain(err))
+                }
+                _ => Err(ResponseReport::with_msg("Failed to make query", resp).into()),
+            }
+        }
 
-    fn handle(self, resp: Response<Vec<u8>>) -> Result<PaginatedQueryOutput<R>> {
-        if resp.status() != StatusCode::OK {
-            return Err(ResponseReport::with_msg("Failed to make query request", &resp).into());
+        _handle_query_response_base(&resp).and_then(|VersionedPaginatedQueryResult::V1(result)| {
+            ClientQueryOutput::try_from(result).map_err(Into::into)
+        })
+    }
+}
+
+/// Different errors as a result of query response handling
+#[derive(Debug)]
+// `Certain` variant is too large (32 bytes), but I think that this enum is not
+// very frequently constructed, so boxing here is unnecessary.
+#[allow(variant_size_differences)]
+pub enum ClientQueryError {
+    /// Specific decoded error from Iroha
+    Certain(QueryError),
```

Comment:
```suggestion
    QueryError(QueryError),
```

Also the opposite of `Indeterminate` isn't `certain`, it's `Determined`.